packages/flutter/lib/src/widgets/router.dart

```diff
@@ -0,0 +1,1327 @@
+
+  /// The back button dispatcher that this object will attempt to take priority
+  /// over when [takePriority] is called.
+  ///
+  /// The parent must have a listener registered before this child object can
+  /// have its [takePriority] or [deferTo] methods used.
+  final BackButtonDispatcher parent;
+
+  /// The parent of this child back button dispatcher decide to let this
+  /// child to handle the invoke the  callback request in
+  /// [BackButtonDispatcher.invokeCallback].
+  ///
+  /// Return a boolean future with true if this child will handle the request;
+  /// otherwise, return a boolean future with false.
+  @protected
+  Future<bool> notifiedByParent(Future<bool> defaultValue) {
+    return invokeCallback(defaultValue);
+  }
+
+  @override
+  void takePriority() {
+    parent.deferTo(this);
+    super.takePriority();
+  }
+
+  @override
+  void deferTo(ChildBackButtonDispatcher child) {
+    assert(hasCallbacks);
+    super.deferTo(child);
+  }
+
+  @override
+  void removeCallback(ValueGetter<Future<bool>> callback) {
+    super.removeCallback(callback);
+    if (!hasCallbacks)
+      parent.forget(this);
+  }
+}
+
+/// A delegate that is used by the [Router] widget to parse a route information
+/// into a configuration of type T.
+///
+/// This delegate is used when the [Router] widget is first built with initial
+/// route information from [Router.routeInformationProvider] and any subsequent
+/// new route notifications from it. The [parseRouteInformation] widget calls
+/// the [parseRouteInformation] with the route information.
+abstract class RouteInformationParser<T> {
+  /// Abstract const constructor. This constructor enables subclasses to provide
+  /// const constructors so that they can be used in const expressions.
+  const RouteInformationParser();
+
+  /// Converts the given route information into parsed data to pass to a
+  /// [RouterDelegate].
+  ///
+  /// The method should return a future which completes when the parsing is
+  /// complete. The parsing may be asynchronous if, e.g., the parser needs to
+  /// communicate with the OEM thread to obtain additional data about the route.
+  ///
+  /// Consider using a [SynchronousFuture] if the result can be computed
+  /// synchronously, so that the [Router] does not need to wait for the next
+  /// microtask to pass the data to the [RouterDelegate].
+  Future<T> parseRouteInformation(RouteInformation routeInformation);
+
+  /// Restore the route information from the given configuration.
+  ///
+  /// This is not required if you do not opt for the route information reporting, which
+  /// is used for updating browser history for the web application. If you
+  /// decides to opt in, you must also overrides this method to return a
+  /// route information.
+  ///
+  /// In practice, the route information returns from this method must be able to
+  /// produce the same configuration when it is parsed by this route information parser
```

Comment:
"The [parseRouteInformation] method must produce an equivalent configuration when passed this method's return value."